improve flex decoding perf by avoid unbalanced splitting

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #170343
* __->__ #167935

Flex decoding previously would do unbalanced split in kv length dimension. That causes many SM having no meaningful work to do. The PR fix that to make the split balanced.

## Accuracy Test
- make sure the answer still make sense from a model: https://gist.github.com/shunting314/9b1861d65d283417c4c5e483b5fd6bb5 
- use vllm attn backends perf comparison script ( https://gist.github.com/shunting314/dbeb935591456b5dafe866787945ce36 ) to explicitly verify the correctness
- CI

## Perf test:
Setup: need to use a small batch size == 4 to make sure we have a single BLOCK_M. Flex decoding does not support multiple BLOCK_M right now.

From e2e result, we only see 2-3% speedup because full cudagraph is not enabled and the job is very overhead bounded. (look into the cudagraphs issue later) But if we zoom into the flex-decoding itself, we clearly see ***6x*** speedup for the kernel we optimized.

before:
<img width="684" height="157" alt="Screenshot 2025-12-11 at 11 17 06 AM" src="https://github.com/user-attachments/assets/e51dd133-5928-4672-b07a-6989ff020460" />

after:
<img width="712" height="149" alt="Screenshot 2025-12-11 at 11 17 53 AM" src="https://github.com/user-attachments/assets/a6d42515-962d-4044-a495-6da220de5c0f" />



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Chillee @drisspg @yanboliang @BoyuanFeng @chenyang78

shunting314 committed:
WIP: flex decoding fix

[ghstack-poisoned]

shunting314 committed:
Update on "WIP: flex decoding fix"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

shunting314 committed:
Update on "WIP: flex decoding improvement"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

shunting314 committed:
Update on "WIP: flex decoding improvement"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

drisspg commented on torch/_inductor/kernel/flex/templates/flex_decode.py.jinja:
```diff
@@ -120,6 +111,19 @@
+
+    # Make sure each split is a multiple of BLOCK_N
+    if HAS_FULL_BLOCKS:
+        TILE_KV_OG = tl.cdiv(KV_LEN, SPLIT_KV)
```
why this for full_blocks?

shunting314 replied:
There is no fundamental reason that the optimization can not be applied to full block. But we don't have a tests covering that due to this setting in vllm: https://github.com/vllm-project/vllm/blob/61249b177de1566027fc74e9b51b45a4c973eb47/vllm/v1/attention/backends/flex_attention.py#L615-L616 .

I think it would be good to revisit if one day vllm can pass in full blocks

drisspg commented on torch/_inductor/kernel/flex/flex_decoding.py:
```diff
@@ -70,16 +84,27 @@ def _use_flex_decoding(query, kv_indices, value, kernel_options, enable_gqa) ->
         and valid_block_mask_num_heads
         and pw_of_two
     )
+    log.debug(
```
noob q does this do anything weird w/ guards? I imagine no right

shunting314 replied:
I don't think so. The flags printed are all bool rather than sympy expressions. Even if they are sympy expressions, printing them should not add guards.

drisspg replied:
Yeah I mean in theory we should probably make this change as we we should forsure have a non zero number of full blocks just need to figure out the fast way to check if the page is full. That is fine can you make a code comment about this

drisspg reviewed (approved):
One comment but looks good!

shunting314 committed:
Update on "improve flex decoding perf by avoid unbalanced splitting"


Flex decoding previously would do unbalanced split in kv length dimension. That causes many SM having no meaningful work to do. The PR fix that to make the split balanced.

## Accuracy Test
- make sure the answer still make sense from a model: https://gist.github.com/shunting314/9b1861d65d283417c4c5e483b5fd6bb5 
- use vllm attn backends perf comparison script ( https://gist.github.com/shunting314/dbeb935591456b5dafe866787945ce36 ) to explicitly verify the correctness
- CI

## Perf test:
Setup: need to use a small batch size == 4 to make sure we have a single BLOCK_M. Flex decoding does not support multiple BLOCK_M right now.

From e2e result, we only see 2-3% speedup because full cudagraph is not enabled and the job is very overhead bounded. (look into the cudagraphs issue later) But if we zoom into the flex-decoding itself, we clearly see ***6x*** speedup for the kernel we optimized.

before:
<img width="684" height="157" alt="Screenshot 2025-12-11 at 11 17 06 AM" src="https://github.com/user-attachments/assets/e51dd133-5928-4672-b07a-6989ff020460" />

after:
<img width="712" height="149" alt="Screenshot 2025-12-11 at 11 17 53 AM" src="https://github.com/user-attachments/assets/a6d42515-962d-4044-a495-6da220de5c0f" />



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Chillee drisspg yanboliang BoyuanFeng chenyang78

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge
